### ISSUE TITLE
Reader event 0 and worker pods evicted within 15 hours of run for medium-scale

### ISSUE THREAD

**Issue by Raghuveera-Rao** [pravega/pravega#4316]
Environment details: PKS / K8 with medium cluster:

```
3 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
5 worker nodes @ xlarge.cpu(8 cpu, 8 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```

Deployment details:
```
bookkeeper:0.6.0-2401.83b62b8
pravega:0.6.0-2401.83b62b8
openjdk:8
pravega/pravega-operator:0.4.2-rc2
pravega/zookeeper:0.2.4
pravega/zookeeper-operator:0.2.4
```

Started longevity medium-scale

I am also running this script inside the worker pod to clear up logs.
```
while true ; do rm -rf controller_server_`date --date="yesterday" +%Y-%m-%d`*; sleep 3600; done  &
```

Pods getting evicted within 15 hours of run
$kubectl get pods -n pravega-longevity
```
NAME                                            READY   STATUS    RESTARTS   AGE
medium-scale-manager-6f69fd6cf7-c5rmp           1/1     Running   0          17h
medium-scale-worker-7d7548dc69-48t22            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-57m4k            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-7vrjg            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-8mc5w            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-b5mxj            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-bj2sj            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-bjrzz            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-fqsl7            1/1     Running   0          179m
medium-scale-worker-7d7548dc69-g674l            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-gh2tc            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-h2bvf            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-j9nd2            0/1     Evicted   0          17h
medium-scale-worker-7d7548dc69-k29mb            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-lhctc            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-lvkpg            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-mlrkp            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-p8975            0/1     Evicted   0          179m
medium-scale-worker-7d7548dc69-pczks            0/1     Evicted   0          179m
```

Readers event is 0
Readers (8/8): `events:0`

```
INFO  [2019-10-23 11:32:22,257] io.pravega.longevity.utils.PerformanceUtils: Readers (8/8): events:0, events/sec:0, KB/sec:0.0
172.28.68.29 - - [23/Oct/2019:11:32:22 +0000] "POST /v1/longevity/performance HTTP/1.1" 200 837 "-" "Jersey/2.25.1 (HttpUrlConnection 1.8.0_131)" 3
INFO  [2019-10-23 11:32:52,255] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 3h 5m
INFO  [2019-10-23 11:32:52,255] io.pravega.longevity.utils.PerformanceUtils: Writers (4/4): events:44,257,300, events/sec:3,987, KB/sec:3056.9229
INFO  [2019-10-23 11:32:52,255] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-10-23 11:32:52,255] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 0, average block time (ms): 0
INFO  [2019-10-23 11:32:52,255] io.pravega.longevity.utils.PerformanceUtils: Readers (8/8): events:0, events/sec:0, KB/sec:0.0
INFO  [2019-10-23 11:32:52,257] io.pravega.longevity.utils.PerformanceUtils: New runtime for b7289442-aec8-4491-bcdc-e2202dbe47e7  Running for 17h 48m
```

Evicted pod description:-
$kubectl describe pod -n pravega-longevity medium-scale-worker-7d7548dc69-48t22
```
Name:           medium-scale-worker-7d7548dc69-48t22
Namespace:      pravega-longevity
Priority:       0
Node:           b7a36e3d-2a90-44ad-889a-7a77dba2f9e6/
Start Time:     Wed, 23 Oct 2019 02:27:41 -0600
Labels:         app=pravega-longevity
                pod-template-hash=7d7548dc69
                test=medium-scale
                type=worker
Annotations:    <none>
Status:         Failed
Reason:         Evicted
Message:        Pod The node had condition: [DiskPressure].
IP:
Controlled By:  ReplicaSet/medium-scale-worker-7d7548dc69
Containers:
  medium-scale-worker:
    Image:      devops-repo.isus.emc.com:8116/nautilus/pravega-longevity:26.5e8ec4b-0.6.0-2401.83b62b8
    Port:       <none>
    Host Port:  <none>
    Command:
      /app/bin/worker
    Environment:
      TEST_MANAGER_HOST:  http://medium-scale-manager:2020
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-rt7zr (ro)
Volumes:
  default-token-rt7zr:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-rt7zr
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

Comeplete k8s logs is very large. Will provide details in jira. Also grafana screenshots will be shared in jira.

**Comment by RaulGracia** [comment]
@Raghuveera-Rao please, attach on each issue not only the node characteristics (which look pretty low in terms of resources), but also the manifest you have used to deploy Pravega (that determines the resource consumption of the system). 

The evictions of longevity worker pods point to a problem of low resources in nodes, based on the workload induced. Please, also post the resources assigned to the longevity pods as well. Note that once a longevity reader is evicted, the code seems not able to handle properly a restart and then it cannot continue reading events. But the root cause here seems to be low resource in the cluster itself, given the workload induced and the Pravega manifest definition.

**Comment by Raghuveera-Rao** [comment]
@RaulGracia  I will attach Pravega manifest in jira correspondence issue...

**Comment by shrids** [comment]
Findings based on the logs of the Oct 31st run.

*Summary*:
This seems more of a resource /cluster related issue here.
The readers lagged and hence faced the `TruncationDataException` since the retention policy of the stream caused truncation of the stream. This is as per the current reader API design.
At a high level observe that there is a general slowdown with this Pravega setup (both the reads and writes have slowed down) and the exception is expected segments the readers are reading from are truncated.
The readers will be able to read the next events post truncation if a new read api is performed ( the test does not do that and it stops for every exception.)

*Details*:

  -   Test started @ `[2019-10-30 14:39:52,489] `

  -   A stream is created which has a retention policy is set for 1 hour.
    ```
    INFO  [2019-10-30 14:39:54,537] io.pravega.client.admin.impl.StreamManagerImpl: Creating scope/stream: longevity/mediumScale with configuration: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=BY_RATE_IN_EVENTS_PER_SEC, targetRate=200, scaleFactor=2, minNumSegments=4), retentionPolicy=RetentionPolicy(retentionType=TIME, retentionParam=3600000), timestampAggregationTimeout=0)
    ```

  -  The test progressed with the readers reading around 4K events/sec and the writers writing 4K events /sec.   Two hours into the run we can observe the writers have slowed down to 2K events/sec while the readers have slowed down to 1.5K events/sec.

    ```
    // 5 mins after the run started.
    INFO  [2019-10-30 14:43:00,371] io.pravega.longevity.utils.PerformanceUtils: Writers (4/4): events:722,133, events/sec:4,011, KB/sec:3075.6865
    INFO  [2019-10-30 14:43:00,371] io.pravega.longevity.utils.PerformanceUtils: Readers (8/8): events:722,133, events/sec:4,011, KB/sec:3075.6865

    // 2 hours after the run started.
    INFO  [2019-10-30 16:35:30,364] io.pravega.longevity.utils.PerformanceUtils: Writers (4/4): events:13,953,015, events/sec:2,013, KB/sec:1543.6846
    INFO  [2019-10-30 16:35:30,364] io.pravega.longevity.utils.PerformanceUtils: Readers (8/8): events:10,006,519, events/sec:1,443, KB/sec:1107.0635
    ```

  -   The readers spawned by the longevity tool observe `TruncationDataException` after 2hours 43 m.

```
 INFO  [2019-10-30 17:23:00,362] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 2h 43m
 INFO  [2019-10-30 17:23:05,727] io.pravega.longevity.testmanager.manager.TestPerformanceCollector: Received Error From task taskID : null
 io.pravega.client.stream.TruncatedDataException
     at io.pravega.client.stream.impl.EventStreamReaderImpl.handleSegmentTruncated(EventStreamReaderImpl.java:307)
 ```
- Out of the 3 segment stores which were part of the setup 2 of them have restarted with OOM.
The version of SSS used is `0.6.0-2401.83b62b8` confirmed the same with Rao, Raghuveera . This does not contain the fix https://github.com/pravega/pravega/pull/4328 which has been merged to the latest 0.6.0 branch.

```
SSS-1 => 
   Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Wed, 30 Oct 2019 08:34:12 -0600
      Finished:     Wed, 30 Oct 2019 20:42:08 -0600

SSS-2 => 
   Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Wed, 30 Oct 2019 08:34:39 -0600
      Finished:     Wed, 30 Oct 2019 19:43:58 -0600
```
- Observing an OOM on the bookie logs too 
```
2019-10-31 01:40:55,982 - ERROR - [bookie-io-1-1:BookieRequestHandler@70] - Unhandled exception occurred in I/O thread or handler
io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 1073741824, max: 1073741824)
    at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:640)
    at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:594)
    at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:764)
```
Note: All the mid-scale worker logs have not been captured as part of the log collection here, we have logs mid-scale worker only from `2019-10-31 04:24:59,750`. Having this would speed up debugging.

Logs: 
[mediumScale_manager.zip](https://github.com/pravega/pravega/files/3819605/mediumScale_manager.zip)

**Comment by shrids** [comment]
Logs:
[pravega-k8s-logging-2019-10-31-00-17-37.zip](https://github.com/pravega/pravega/files/3819620/pravega-k8s-logging-2019-10-31-00-17-37.zip)
